driver-kotlin-sync/src/test/kotlin/com/mongodb/kotlin/client/MongoChangeStreamCursorTest.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.kotlin.client
+
+import com.mongodb.ServerAddress
+import com.mongodb.ServerCursor
+import com.mongodb.client.MongoChangeStreamCursor as JMongoChangeStreamCursor
+import kotlin.reflect.full.declaredFunctions
+import kotlin.reflect.full.declaredMemberProperties
+import kotlin.test.assertEquals
+import org.bson.Document
+import org.junit.jupiter.api.Test
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.verifyNoMoreInteractions
+import org.mockito.kotlin.whenever
+
+class MongoChangeStreamCursorTest {
+    @Test
+    fun shouldHaveTheSameMethods() {
+        val jMongoChangeStreamCursorFunctions =
+            JMongoChangeStreamCursor::class.declaredFunctions.map { it.name }.toSet()
+        val kMongoChangeStreamCursorFunctions =
+            MongoChangeStreamCursor::class.declaredFunctions.map { it.name }.toSet() +
+                MongoChangeStreamCursor::class
+                    .declaredMemberProperties
+                    .filterNot { it.name == "wrapped" }
+                    .map { "get${it.name.replaceFirstChar{c -> c.uppercaseChar() }}" }
```

Comment:
Good question. In Kotlin classes can have both properties and functions.

In certain places I used properties for any getters that are immutable and don't change state.

Officially the [coding convention](https://kotlinlang.org/docs/coding-conventions.html#functions-vs-properties) states:

> **Functions vs properties**﻿
> In some cases functions with no arguments might be interchangeable with read-only properties. Although the semantics are similar, there are some stylistic conventions on when to prefer one to another.
>
> Prefer a property over a function when the underlying algorithm:
> 
> - does not throw
> - is cheap to calculate (or cached on the first run)
> - returns the same result over invocations if the object state hasn't changed

So for `MongoChangeStreamCursor` I used a property for `ServerAddress` but functions for `getResumeToken` as that changes state.

I could use properties for more functions that meet the above criteria but I think functions signalling mutable property values and properties signalling immutable values is the right decision.  As this is a wrapper it makes no sense to add a setter for a property (which would declare it mutable) as that is all internal to the wrapper.

@cmelchior - as my goto Kotlin expert - is my interpretation of the use of properties idiomatic? Or Should I be using properties more?

Reply:
(Consider resolved from my side, pending @cmelchior response.)

Reply:
I would partly agree and partly disagree. We are using the same reference in their coding convention but probably lean slightly more towards exposing them as properties since that is easier to consume when writing the code and relying on autocomplete.

Their heuristics are also a bit elastic, i.e. what is "cheap" and what determines "object state".

I think it is a mistake to treat properties as constants though, both `val` and `var` properties are allowed to change.

In this particular case, I would probably lean towards turning `getResumeToken` into a `resumeToken` property.

Some examples from the Realm Kotlin SDK:

- [`User`](https://github.com/realm/realm-kotlin/blob/main/packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/User.kt#L38) .. most things are properties, even things that "change".
- [`BaseRealm`](https://github.com/realm/realm-kotlin/blob/main/packages/library-base/src/commonMain/kotlin/io/realm/kotlin/BaseRealm.kt#L33) ... we had a lengthy discussion about `schema()` but it ended up a function because it is an expensive data structure to build and return.

API design is tricky :D

Reply:
Sounds good - updated a few functions to properties